src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,689 @@
+of_int_seq (nth_hom f r) (is_cau_seq_nth_hom f_compat r)
+
+lemma lim_nth_hom_spec (r : R) :
+  ∀ ε : ℝ, 0 < ε → ∃ N : ℕ, ∀ n ≥ N, ∥lim_nth_hom f_compat r - nth_hom f r n∥ < ε :=
+begin
+  intros ε hε,
+  obtain ⟨ε', hε'0, hε'⟩ : ∃ v : ℚ, (0 : ℝ) < v ∧ ↑v < ε := exists_rat_btwn hε,
+  norm_cast at hε'0,
+  obtain ⟨N, hN⟩ := padic_norm_e.defn (nth_hom_seq f_compat r) hε'0,
+  use N,
+  intros n hn,
+  apply lt.trans _ hε',
+  change ↑(padic_norm_e _) < _,
+  norm_cast,
+  convert hN _ hn,
+  simp [nth_hom, lim_nth_hom, nth_hom_seq, of_int_seq],
+end
+
+lemma lim_nth_hom_zero : lim_nth_hom f_compat 0 = 0 :=
+by simp [lim_nth_hom]; refl
+
+lemma lim_nth_hom_one : lim_nth_hom f_compat 1 = 1 :=
+subtype.ext $ quot.sound $ nth_hom_seq_one _
+
+lemma lim_nth_hom_add (r s : R) : lim_nth_hom f_compat (r + s) = lim_nth_hom f_compat r + lim_nth_hom f_compat s :=
+subtype.ext $ quot.sound $ nth_hom_seq_add _ _ _
+
+lemma lim_nth_hom_mul (r s : R) : lim_nth_hom f_compat (r * s) = lim_nth_hom f_compat r * lim_nth_hom f_compat s :=
+subtype.ext $ quot.sound $ nth_hom_seq_mul _ _ _
+
+-- TODO: generalize this to arbitrary complete discrete valuation rings
+/--
+`lift f_compat` is the limit of a sequence `f` of compatible ring homs `R →+* zmod (p^k)`,
+with the equality `lift f_compat r = `padic_int.lim_nth_hom f_compat r`.
+-/
+def lift : R →+* ℤ_[p] :=
+{ to_fun := lim_nth_hom f_compat,
+  map_one' := lim_nth_hom_one f_compat,
+  map_mul' := lim_nth_hom_mul f_compat,
+  map_zero' := lim_nth_hom_zero f_compat,
+  map_add' := lim_nth_hom_add f_compat }
+
+omit f_compat
+
+lemma lift_sub_val_mem_span (r : R) (n : ℕ) :
+  (lift f_compat r - (f n r).val) ∈ (ideal.span {↑p ^ n} : ideal ℤ_[p]) :=
+begin
+  obtain ⟨k, hk⟩ := lim_nth_hom_spec f_compat r _
+    (show (0 : ℝ) < p ^ (-n : ℤ), from nat.fpow_pos_of_pos hp_prime.pos _),
+  have := le_of_lt (hk (max n k) (le_max_right _ _)),
+  rw norm_le_pow_iff_mem_span_pow at this,
+  dsimp [lift],
+  rw sub_eq_sub_add_sub (lim_nth_hom f_compat r) _ ↑(nth_hom f r (max n k)),
+  apply ideal.add_mem _ _ this,
+  rw [ideal.mem_span_singleton],
+  simpa only [ring_hom.eq_int_cast, ring_hom.map_pow, int.cast_sub] using
+    (int.cast_ring_hom ℤ_[p]).map_dvd (pow_dvd_nth_hom_sub f_compat r n (max n k) (le_max_left _ _)),
+end
+
+/--
+One part of the universal property of `ℤ_[p]` as a projective limit.
+See also `padic_int.lift_unique`.
+-/
+lemma lift_spec (n : ℕ) : (to_zmod_pow n).comp (lift f_compat) = f n :=
+begin
+  ext r,
+  haveI : fact (0 < p ^ n) := nat.pow_pos (nat.prime.pos ‹_›) n,
+  rw [ring_hom.comp_apply, ← zmod.cast_val (f n r), ← (to_zmod_pow n).map_nat_cast,
+      ← sub_eq_zero, ← ring_hom.map_sub, ← ring_hom.mem_ker, ker_to_zmod_pow],
+  apply lift_sub_val_mem_span,
+end
+
+/--
+One part of the universal property of `ℤ_[p]` as a projective limit.
+See also `padic_int.lift_spec`.
+-/
+lemma lift_unique (g : R →+* ℤ_[p]) (hg : ∀ n, (to_zmod_pow n).comp g = f n) :
+  lift f_compat = g :=
```

Comment:
Could you also prove that two maps to Z_p which agree mod p^n for all n agree?

Do we also have that x and y in Z_p are equal iff their reductions mod p^n are equal for all n?

Reply:
> Do we also have that x and y in Z_p are equal iff their reductions mod p^n are equal for all n?

This is `ext_to_of_zmod_pow` at the bottom of the file, right? (At least the nontrivial direction of the iff.)

> Could you also prove that two maps to Z_p which agree mod p^n for all n agree?

Hmm, I think the infrastructure is there, but I'm not sure how much extra work it would take. @jcommelin ?

Reply:
The two maps are both equal to the lift, by this `lift_unique` lemma.

Reply:
Ah, yeah -- I had misinterpreted Kevin's comment. I've added `to_zmod_pow_eq_iff_ext` at the end of the file but I'm not sure about the name...